be/src/storage/binlog_reader.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/binlog_reader.h"
+
+#include <utility>
+
+#include "storage/binlog_manager.h"
+#include "storage/rowset/rowid_range_option.h"
+#include "storage/rowset/segment_options.h"
+
+namespace starrocks {
+
+BinlogReader::BinlogReader(int64_t reader_id, std::shared_ptr<BinlogManager> binlog_manager, vectorized::VectorizedSchema& schema, int chunk_size)
+        : _reader_id(reader_id),
+          _binlog_manager(std::move(binlog_manager)),
+          _schema(std::move(schema)),
+          _chunk_size(chunk_size) {}
+
+BinlogReader::~BinlogReader() {
+    _reset();
+}
+
+Status BinlogReader::seek(int64_t version, int64_t seq_id) {
+    if (version < _next_version || (version == _next_version && seq_id > _next_seq_id)) {
+        return Status::InternalError(strings::Substitute("Binlog can only be read forward, next position"
+            " <$0, $1>, seek to <$2, $3>", _next_version, _next_seq_id, version, seq_id));
+    }
+
+    if (_next_version == version && _next_seq_id == seq_id) {
+        return Status::OK();
+    }
+
+    RETURN_IF_ERROR(_seek_to_file_meta(version, seq_id));
+    RETURN_IF_ERROR(_seek_to_segment_row(seq_id));
+    LOG(INFO) << "Binlog reader " << _reader_id << ", seek to version " << version << ", seq_id " << seq_id;
+    return Status::OK();
+}
+
+Status BinlogReader::get_next(vectorized::ChunkPtr* chunk, int64_t max_version_exclusive) {
+    if (_next_version >= max_version_exclusive) {
+        return Status::EndOfFile("End of max version " + max_version_exclusive);
+    }
+
+    chunk->get()->reset();
+    if (_log_entry_info->log_entry->entry_type() == EMPTY) {
+        _next_version += 1;
+        _next_seq_id = 0;
+        return Status::OK();
+    }
+
+    Status status = _segment_iterator->get_next(chunk->get());
+    if (!status.ok() && !status.is_end_of_file()) {
+        return status;
+    }
+
+    // TODO append binlog meta columns
```

Comment:
How to add binlog meta?  Does BinlogReader support `filter`?

This maybe a problem, or `BinlogReader` just not support filter, always output full schema?